ui/raidboss/data/06-ew/ultimate/dragonsongs_reprise_ultimate.ts

```diff
@@ -818,14 +827,306 @@ const triggerSet: TriggerSet<Data> = {
+      },
+    },
+    {
+      id: 'DSR Dive From Grace Tower 2 and 3',
+      // 670E Dark High Jump
+      // 670F Dark Spineshatter Dive
+      // 6710 Dark Elusive Jump
+      // Defaults: (as players will be coming from stack)
+      //   Spineshatter Right, Elusive Left, All Face East
+      //   High Jump North if solo, no assignment if all circle
+      //   2s Southeast/Southwest, no assignment if circle
+      //   Assumes North Party Stack
+      type: 'Ability',
+      netRegex: NetRegexes.ability({ id: ['670E', '670F', '6710'], source: 'Nidhogg', capture: false }),
+      preRun: (data) => data.diveFromGraceTowerCounter = (data.diveFromGraceTowerCounter ?? 0) + 1,
+      delaySeconds: 0.1,
+      suppressSeconds: 1,
+      infoText: (data, _matches, output) => {
+        const num = data.diveFromGraceNum[data.me];
+        if (!num) {
+          console.error(`DFG Tower 2 and 3: missing number: ${JSON.stringify(data.diveFromGraceNum)}`);
+          return;
+        }
+
+        // With two Cartesian Coordinates, calculate distance
+        const distance = (a?: number[], b?: number[]) => {
+          if (
+            a === undefined || a[0] === undefined || a[1] === undefined ||
+            b === undefined || b[0] === undefined || b[1] === undefined
+          ) {
+            console.error(`DFG Tower 2 and 3: Missing targetX, targetY`);
+            return -1;
+          }
+          const x = a[0] - b[0];
+          const y = a[1] - b[1];
+          return Math.sqrt(x * x + y * y);
+        };
+
+        // Map position values and player name keys to arrays
+        let [posA, posB, posC] = Object.values(data.diveFromGracePositions);
+        let [nameA, nameB, nameC] = Object.keys(data.diveFromGracePositions);
+
+        // If undefined position, will not be able to predict position
+        if (posA === undefined)
+          posA = [];
+        if (posB === undefined)
+          posB = [];
+        if (posC === undefined)
+          posC = [];
+        const posAX = posA[0] ?? 0;
+        const posBX = posB[0] ?? 0;
+        const posCX = posC[0] ?? 0;
+
+        // If undefined keys, map to non-player names
+        if (nameA === undefined)
+          nameA = '???';
+        if (nameB === undefined)
+          nameB = '???';
+        if (nameC === undefined)
+          nameC = '???';
+
+        // Dive 1 and Dive 3 have 3 players
+        if (data.diveFromGraceTowerCounter !== 2) {
+          // Get Distances of assumed "Triangle" pattern created by 3 players
+          const distanceAB = distance(posA, posB);
+          const distanceAC = distance(posA, posC);
+          const distanceBC = distance(posB, posC);
+          const distances = [distanceAB, distanceAC, distanceBC];
+
+          // Sort the distances, largest to smallest
+          const sorted = distances.sort((a, b) => b - a);
+
+          // Assuming the two players furthest apart are the left/right players
```

Comment:
Sorry for misunderstanding, but why look at distance vs using the x coordinate to figure out left vs right?

Reply:
The distance is actually just to get the circle player's position for the 1's and 3's. At the time it seemed like a simple solution to handle orientations that were of any cardinal as I was not sure if the stack was required to be in a specific cardinal. Any cardinal such as ENW, NWS, ESW, NSE.

Reply:
There was a mistake in one of the comparisons.

```
          const distanceAB = distance(posA, posB);
          const distanceAC = distance(posA, posC);
          const distanceBC = distance(posB, posC);
          const distances = [distanceAB, distanceAC, distanceBC];
```
These will give us the distances between players, which is important because by sorting distances and then looking back by comparing distanceAB, distanceAC, and distanceBC to the sorted values, we can determine which players were left/right and south player through elimination, assuming a triangle-like pattern.

Reply:
I guess I still don't understand why distances need to be used here.  It seems like you could be able to sort based on x coordinate, and then there would be a lot less logic here.

Reply:
I think pretty much the same thing can be done with sorting posX and taking the middle value to get who was middle, which will work for n/s.